### PR TITLE
enable parallel rustc front end in nightly builds

### PR DESCRIPTION
**Pull request by SparrowLii**
Refers to the [MCP](https://github.com/rust-lang/compiler-team/issues/681), this pr does:
1. Enable the parallel front end in nightly builds, and keep the default number of threads as 1. Then users can use the parallel rustc front end via -Z threads=n option.

2. Set it up to serial front end for beta/stable builds via bootstrap.

3. Switch over the alt builders from parallel rustc to serial, so we have artifacts without parallel to test against the artifacts with parallel.

r? @oli-obk 

cc @cjgillot @nnethercote @bjorn3 @Kobzol

### CODE REVIEW

File: src/bootstrap/src/core/config/config.rs [hunk 1]
```diff
@@ -1429,7 +1430,9 @@ impl Config {
             set(&mut config.use_lld, rust.use_lld);
             set(&mut config.lld_enabled, rust.lld);
             set(&mut config.llvm_tools_enabled, rust.llvm_tools);
-            config.rustc_parallel = rust.parallel_compiler.unwrap_or(false);
+            config.rustc_parallel = rust
+                .parallel_compiler
+                .unwrap_or(config.channel == "dev" || config.channel == "nightly");
```
Comment (bjorn3) [inline src/bootstrap/src/core/config/config.rs:1435]: Can you reflect this change in `config.example.toml` too to avoid confusion?

**Review by oli-obk — APPROVED** [review]
I'm gonna let people chime in on this for a day or two. If I don't approve by Thursday myself, r=me

**Comment by lqd** [comment]
Do we need to land https://github.com/rust-lang/rust/pull/115220 prior to this PR?

**Comment by oli-obk** [comment]
More than one thread is deadlock prone and anyone opting into it with the -Z flag has to accept that I think

**Comment by SparrowLii** [comment]
@oli-obk I think we can merge now

@bors r=oli-obk

File: src/ci/run.sh [hunk 2]
```diff
@@ -98,8 +98,8 @@ if [ "$DEPLOY$DEPLOY_ALT" = "1" ]; then
   if [ "$NO_LLVM_ASSERTIONS" = "1" ]; then
     RUST_CONFIGURE_ARGS="$RUST_CONFIGURE_ARGS --disable-llvm-assertions"
   elif [ "$DEPLOY_ALT" != "" ]; then
-    if [ "$NO_PARALLEL_COMPILER" = "" ]; then
-      RUST_CONFIGURE_ARGS="$RUST_CONFIGURE_ARGS --set rust.parallel-compiler"
+    if [ "$ALT_PARALLEL_COMPILER" = "" ]; then
+      RUST_CONFIGURE_ARGS="$RUST_CONFIGURE_ARGS --set rust.parallel-compiler=false"
```
Comment (klensy) [inline src/ci/run.sh:102]: Can't find `NO_PARALLEL_COMPILER` or `ALT_PARALLEL_COMPILER` in codebase, is there some secret?

Comment (SparrowLii) [reply src/ci/run.sh:102]: I think it's from env variable

**Review by davidtwco — APPROVED** [review]
Dependency changes are okay 👍

**Comment by SparrowLii** [comment]
Powerpc and mips do not support AtomicU64 (mainly used to assign `QueryJobId`s for every query during compilation). We can use the `portable-atomic` crate for these two target, requiring adding a new rustc's dependency though. 

@oli-obk Could you review the PR again?

**Comment by oli-obk** [comment]
r? compiler I am on leave

File: src/ci/run.sh (hunk 2 above)
Comment (klensy) [reply src/ci/run.sh:102]: Envs checked there, but where they set, i can't find.

**Comment by pnkfelix** [comment]
* I am not sure that wg-parallel-rustc was anticipating there being this much of a regression here.
* asked for followup in [zulip for wg-parallel-rustc](https://rust-lang.zulipchat.com/#narrow/stream/187679-t-compiler.2Fwg-parallel-rustc/topic/performance.20results/near/400790214)
* not marking as triaged

**Comment by SparrowLii** [comment]
@pnkfelix Sorry for not explaining in the MCP. We [decided](https://rust-lang.zulipchat.com/#narrow/stream/187679-t-compiler.2Fwg-parallel-rustc/topic/meeting.202023-02-27/near/338283965) in the working group meeting early this year that wall-time is a better choice to measuring the perf result of parallel front end. Therefore, the measurement we choose is wall-time+non-relevent. (You may also need to exclude secondary Categories). In this case, the regression of #117435 should be [1.65%](https://perf.rust-lang.org/compare.html?start=fcca978aa0c9a7443649305dd0ed48c4691a1e89&end=f9b644636f8c74d1d9a3283dc65a6a6f02a97718&stat=wall-time&nonRelevant=true)

**Comment by pnkfelix** [comment]
* after followup in [zulip for wg-parallel-rustc](https://rust-lang.zulipchat.com/#narrow/stream/187679-t-compiler.2Fwg-parallel-rustc/topic/performance.20results/near/400790214), they clarified that their expectation of a 1.65% regression was based on wall-time that includes so-called "non-relevant results". (Our default rustc-perf presentation filters out the non-relevant results, which will boost outliers and thus inflate the true mean effect.)
* concluded that this was anticipated, and marking as triaged

@rustbot label: +perf-regression-triaged

**Comment by SparrowLii** [comment]
Thanks!

File: src/bootstrap/src/lib.rs [hunk 3]
```diff
@@ -77,7 +77,7 @@ const LLD_FILE_NAMES: &[&str] = &["ld.lld", "ld64.lld", "lld-link", "wasm-ld"];
 ///
 /// If you make any major changes (such as adding new values or changing default values), please
 /// ensure that the associated PR ID is added to the end of this list.
-pub const CONFIG_CHANGE_HISTORY: &[usize] = &[115898, 116998];
+pub const CONFIG_CHANGE_HISTORY: &[usize] = &[115898, 116998, 117435];
```
Comment (RalfJung) [inline src/bootstrap/src/lib.rs:80]: When you do this, please explain in the PR description what we have to do to update our configs. I was sent here by bootstrap and now I have no idea what I'm supposed to do.

**Pull request merged**